src/Datadog.Trace/TraceContext.cs

```diff
@@ -105,6 +105,7 @@ public void CloseSpan(Span span)
                 if (_openSpans == 0)
                 {
                     Tracer.Write(_spans);
+                    _spans = new List<Span>();
```

Comment:
The actual change. This is done while still holding the `_lock`

Reply:
Actually, can we make a copy and pass that into Tracer.Write?
Then we can release the lock as soon as _spans is set to a new instance.
```
var copy = _spans;
_spans = new List<Span>();
// release lock
Tracer.Write(copy);

Reply:
Hmm. At first I thought this was a no-brainer, but I don't think there's too much benefit to move some logic outside of the lock because our `openSpans` counter checks and modifications should all be done inside the lock. And the Write call only pushes the given spans onto a stack (it doesn't block the thread to perform a potentially long write) so the current implementation is creating too large an overhead. Would you be fine with leaving as-is?

Reply:
It does lock per push within the call, and you're right, it's just an in-memory operation. An in-memory operation can still take some time depending on count.. 
The typical use of a lock is to hold it for only as long as you need to, so I suppose this is a bit of defensive programming. Defending us against the potential change of the `Tracer.Write` implementation.
That being said, we are planning to re-work this whole buffering system, the question is just about when we actually get to it. 
The zero check will likely go away if we move to a streaming model.

Reply:
That said, it won't make much of a difference today, it's your call.

Reply:
You're absolutely right about trying to minimize the operations inside lock, and it's an easy change to make that only has upside. I made the suggested change in my latest commits.